File: src/main/java/org/hydev/mcpm/client/search/SearcherFactory.java
package org.hydev.mcpm.client.search;

/**
 * Factory that produces Searcher objects based on the input type.
 *
 */
public class SearcherFactory {
    SearcherByName nameSearcher;
    SearcherByKeyword keywordSearcher;
    SearcherByCommand commandSearcher;

    /**
     * Returns the new searcher object based on the input type.
     *
     * @param input Contains the search type in particular. See SearchPackagesInput for details.
     */
    public Searcher createSearcher(SearchPackagesInput input) {
        return switch (input.type()) {
            case BY_NAME ->
                nameSearcher == null ?
                    nameSearcher = new SearcherByName() : nameSearcher;

            case BY_COMMAND ->
                commandSearcher == null ?
                        commandSearcher = new SearcherByCommand() : commandSearcher;

            case BY_KEYWORD ->
                keywordSearcher == null ?
                        keywordSearcher = new SearcherByKeyword() : keywordSearcher;
        };
    }
}

File: src/main/java/org/hydev/mcpm/client/search/SearcherFactory.kt
package org.hydev.mcpm.client.search

import org.hydev.mcpm.client.search.SearchPackagesType.*

/**
 * Factory that produces Searcher objects based on the input type.
 *
 */
class SearcherFactory
{
    private val nameSearcher: SearcherByName by lazy { SearcherByName() }
    private val keywordSearcher: SearcherByKeyword by lazy { SearcherByKeyword() }
    private val commandSearcher: SearcherByCommand by lazy { SearcherByCommand() }
    private val idSearcher: SearcherById by lazy { SearcherById() }

    private val map = mapOf(BY_NAME to nameSearcher, BY_COMMAND to commandSearcher,
        BY_KEYWORD to keywordSearcher, BY_ID to idSearcher)

    /**
     * Returns the new searcher object based on the input type.
     *
     * @param input Contains the search type in particular. See SearchPackagesInput for details.
     */
    fun createSearcher(input: SearchPackagesInput) = map[input.type]
}
